plugins/ep963x/fu-ep963x-device.c

```diff
@@ -0,0 +1,347 @@
+/*#
+ * Copyright (C) 2020 Richard Hughes <richard@hughsie.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include "fu-chunk.h"
+#include "fu-ep963x-common.h"
+#include "fu-ep963x-device.h"
+#include "fu-ep963x-firmware.h"
+
+struct _FuEp963xDevice {
+	FuHidDevice		 parent_instance;
+};
+
+G_DEFINE_TYPE (FuEp963xDevice, fu_ep963x_device, FU_TYPE_HID_DEVICE)
+
+#define FU_EP963_DEVICE_TIMEOUT			5000	/* ms */
+
+static gboolean
+fu_ep963x_device_write (FuEp963xDevice *self,
+			guint8 cmd, guint8 *buf, gsize bufsz,
+			GError **error)
+{
+	guint8 bufhw[FU_EP963_FEATURE_ID1_SIZE] = {
+		FU_EP963_USB_CONTROL_ID, cmd, 0x0,
+	};
+	if (buf != NULL) {
+		if (!fu_memcpy_safe (bufhw, sizeof(bufhw), 0x02,	/* dst */
+				     buf, bufsz, 0x0,			/* src */
+				     bufsz, error))
+			return FALSE;
+	}
+	return fu_hid_device_set_report (FU_HID_DEVICE (self),
+					 0x00,
+					 bufhw, sizeof(bufhw),
+					 FU_EP963_DEVICE_TIMEOUT,
+					 FU_HID_DEVICE_FLAG_IS_FEATURE,
+					 error);
+}
+
+static gboolean
+fu_ep963x_device_write_icp (FuEp963xDevice *self,
+			    guint8 cmd, guint8 *buf, gsize bufsz,
+			    GError **error)
+{
+	guint8 bufhw[FU_EP963_FEATURE_ID1_SIZE] = { 0x0 };
+
+	/* optional data */
+	if (buf != NULL) {
+		if (!fu_memcpy_safe (bufhw, sizeof(bufhw), 0x02,	/* dst */
+				     buf, bufsz, 0x0,			/* src */
+				     bufsz, error))
+			return FALSE;
+	}
+
+	/* wait for hardware */
+	for (guint i = 0; i < 5; i++) {
+		bufhw[0] = FU_EP963_FEATURE_ID1_SIZE;
+		bufhw[1] = cmd;
+		if (!fu_hid_device_set_report (FU_HID_DEVICE (self), 0x00,
+					       bufhw, sizeof(bufhw),
+					       FU_EP963_DEVICE_TIMEOUT,
+					       FU_HID_DEVICE_FLAG_IS_FEATURE,
+					       error)) {
+			return FALSE;
+		}
+		g_usleep (100 * 1000);
```

Comment:
If all commands are going to have this settle/sleep after, maybe split to a separate function

Reply:
Not sure I follow, can you elaborate a bit pls.

Reply:
It seems with set commands you're sleeping each time after for a settle time period?

Reply:
I think I know what you mean, and I think it uncovered a missing sleep, so nice one there. Fixed.

Reply:
Yeah you adjusted it how I was thinking.  Glad you could make sense of what I meant :)